### ISSUE TITLE
Bring the Async Rust experience closer to parity with sync Rust

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust-project-goals#105]
| Metadata         | |
| --------         | --- |
| Point of contact | @tmandry |
| Team(s)          | [compiler](http://github.com/rust-lang/compiler-team), [lang](http://github.com/rust-lang/lang-team), [libs-api](https://www.rust-lang.org/governance/teams), [spec](https://github.com/rust-lang/spec), [types](https://github.com/rust-lang/types-team) |
| Goal document    | [2025h1/async](https://rust-lang.github.io/rust-project-goals/2025h1/async.html) |

## Summary

Over the next six months, we will continue bringing Async Rust up to par with "sync Rust" by doing the following:

* Telling a complete story for the use of async fn in traits, unblocking wide ecosystem adoption,
* Improving the ergonomics of `Pin`, which is frequently used in low-level async code, and
* Preparing to support asynchronous (and synchronous) generators in the language.

## Why this goal?

This work continues our drive to improve support for async programming in Rust. In 2024H2 we stabilized async closures; explored the generator design space; and began work on the `dynosaur` crate, an experimental proc-macro to provide dynamic dispatch for async functions in traits. In 2025H1 [our plan](https://rust-lang.github.io/rust-project-goals/2025h1/async.html) is to deliver (1) improved support for async-fn-in-traits, completely subsuming the functionality of the [`async-trait` crate](https://crates.io/crates/async-trait); (2) progress towards sync and async generators, simplifying the creation of iterators and async data streams; (3) and improve the ergonomics of `Pin`, making lower-level async coding more approachable. These items together start to unblock the creation of the next generation of async libraries in the wider ecosystem, as progress there has been blocked on a stable solution for async traits and streams.

### Overall program management
* [ ] AFIT story blog post (@tmandry)
### Return type notation (https://github.com/rust-lang/rust/issues/109417)
* [x] Initial implementation (@compiler-errors)
* [x] Author RFC (@nikomatsakis)
* [x] RFC decision ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [x] Finished implementation (@compiler-errors)
* [x] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [x] Standard reviews ([types](https://github.com/rust-lang/types-team), [compiler](http://github.com/rust-lang/compiler-team) ![Team][])
* [x] Author stabilization report (@compiler-errors)
* [ ] Author specification 1st draft (@compiler-errors, @tmandry, @nikomatsakis)
* [ ] Finalize specification text ([spec](https://github.com/rust-lang/spec) ![Team][])
* [ ] Stabilization decision ([lang](http://github.com/rust-lang/lang-team), [types](https://github.com/rust-lang/types-team) ![Team][])
### Unsafe binders (https://github.com/rust-lang/rust/issues/130516)
* [ ] Initial implementation (@compiler-errors)
* [ ] Author RFC (@nikomatsakis)
* [ ] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] RFC decision ([lang](http://github.com/rust-lang/lang-team), [types](https://github.com/rust-lang/types-team) ![Team][])
### Implementable trait aliases
* [ ] Author RFC (@tmandry)
* [ ] Implementation (@compiler-errors)
* [ ] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] Standard reviews ([types](https://github.com/rust-lang/types-team), [compiler](http://github.com/rust-lang/compiler-team) ![Team][])
* [ ] RFC decision ([lang](http://github.com/rust-lang/lang-team), [types](https://github.com/rust-lang/types-team) ![Team][])
### `async fn` in `dyn Trait` (https://github.com/rust-lang/rust/issues/133119)
* [ ] Lang-team experiment (@nikomatsakis)
* [x] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [ ] Implementation (@compiler-errors)
### Pin ergonomics
* [ ] Implementation (@eholk)
* [ ] Author RFC (@eholk)
* [x] Lang-team experiment ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [x] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
### Trait for generators (sync)
* [ ] Implementation (@eholk)
* [ ] Author RFC
* [ ] RFC decision ([libs-api](https://www.rust-lang.org/governance/teams), [lang](http://github.com/rust-lang/lang-team) ![Team][])
* [x] Design meeting ([lang](http://github.com/rust-lang/lang-team) ![Team][])
* [x] Lang-team champion ([lang](http://github.com/rust-lang/lang-team) ![Team][])
### Dynosaur 1.0
* [x] Implementation (@spastorino)
* [x] Standard reviews (@tmandry)

[Team]: https://img.shields.io/badge/Team%20ask-red

**Comment by nikomatsakis** [comment]
This issue is intended for status updates only.

For general questions or comments, please contact the owner(s) directly.

**Comment by tmandry** [comment]
## Status update: July 30, 2024

### Key developments

* Our project goal was merged :)
* Return type notation RFC was merged (rust-lang/rfcs#3654)
* Async closures RFC is in final comment period (rust-lang/rfcs#3668)

### Blockers

We are able to make some progress in each area, but the biggest upcoming blocker we can see is:

* Lang team deciding on final syntax for async closure bounds

### Help wanted

There will be a call for testing async closures on the Inside Rust blog soon. Please check it out and give them a try in your own async code!

We'd like to see an updated RFC for implementable trait aliases that follows the guidelines laid out in [this comment](https://github.com/rust-lang/rfcs/pull/3437#issuecomment-2207205072).

**Comment by nikomatsakis** [comment]
### Executive summary.

We are generally on track with our marquee features: (1) Support for *async closures* is available on Nightly and the lang team arrived at a tentative consensus to keep the existing syntax (written rationale and formal decision are in progress). We issued a [call for testing](https://blog.rust-lang.org/inside-rust/2024/08/09/async-closures-call-for-testing.html) as well which has so far uncovered no issues. (2) Partial support for *Return Type Notation* is available on Nightly with the remainder under review. In addition, *dynamic dispatch for async functions* and *experimental async drop work* both made implementation progress. Async WG reorganization has made no progress.

### Detailed updates

For this year's goal we are focused on three items:

[RFC #3668]: https://github.com/rust-lang/rfcs/pull/3668
[RFC #3654]: https://github.com/rust-lang/rfcs/pull/3654

* Stabilizing async closures:
    * The [Async Closures RFC][RFC #3668] has been accepted and the feature is available for use on Nightly under the `async_closure` feature gate. We put out a [call for testing](https://blog.rust-lang.org/inside-rust/2024/08/09/async-closures-call-for-testing.html); so far, no problems have been reported. Stabilization stalled because of a need to [resolve the syntax](https://github.com/rust-lang/rust/issues/128129). Lang team has reached a tentative internal consensus in favor of the syntax as implemented (`impl async Fn`). Next steps are (1) lang team to write up rationale and finalizing syntax decision; (2) author stabilization report.
* Resolving "Send Bounds" via Return Type Notation:
    * The [Return Type Notation RFC][RFC #3654] has been accepted and partially implemented under the feature gate `return_type_notation`. Bounds like `Trait<method(): Send>` are available on Nightly; bounds like `Trait::method(): Send` will be available once PR [#129629] is merged. Next steps are to (1) merge PR [#129629]; (2) call for testing; and (3) author stabilization report.
* Dyn dispatch for async functions in traits:
    * Async functions in traits were stabilized last year without support for dynamic dispatch. For this year we plan to release a procedural macro that can help users workaround this limitation; native support is planned for future years. The all important task of selecting a crate name was completed, with `dynosaur` being chosen. Basic support is now implemented ([see this example](https://github.com/spastorino/dynosaur/blob/54bdea07a9892670425385db1a239f81fbad939a/dynosaur/examples/next.rs#L1-L5)). Next steps are polish and putting out a call for testing.
* Async drop experiments:
    * PRs with experimental implementation work for async drop are ready and awaiting review ([PR list](https://github.com/rust-lang/rust/pulls?q=is%3Aopen+is%3Apr+author%3Aazhogin+async+drop)).

We have not made progress on Async WG reorganization.

**Comment by nikomatsakis** [comment]
Updates:

* Stabilizing async closures:
    * Lang team discussed syntax and reached preliminary consensus on `async Fn`, subject to also agreeing to include some of "async type" syntax. The rationale was documented in rust-lang/rfcs#3710, which is now open and receiving feedback.
    * Next steps are for lang team to accept the RFC and to author stabilization report.
* Resolving "Send Bounds" via Return Type Notation: 
    * A [call for testing](https://blog.rust-lang.org/inside-rust/2024/09/26/rtn-call-for-testing.html) of RTN was issued and PR https://github.com/rust-lang/rust/pull/129629 was merged.
    * Next steps is to author a stabilization report.
* Dyn dispatch for async functions in traits: 
    * Preliminary implementation work on the `dynosaur` crate was done. Polish and public call for testing remain the next steps.
* Async drop experiments
    * There is a PR with pending work awaiting review. @nikomatsakis has agreed to review.
* Async WG reorganization.
    * No progress. A meeting was scheduled but deferred.

**Comment by tmandry** [comment]
> Preliminary implementation work on the dynosaur crate was done. Polish and public call for testing remain the next steps.

The first version of dynosaur was published: https://docs.rs/dynosaur/latest/dynosaur/attr.dynosaur.html

**Comment by nikomatsakis** [comment]
Updates:

* Stabilizing async closures: The lang team held a [design meeting](https://hackmd.io/w2-etPU7RouzxJ6GJ1fbSg) reviewing the Flavor RFC. Conclusion was that @nikomatsakis will make some edits clarifying the RFC's limited intent (e.g., that is defines a namespacing mechanism and not something more involved) and weakening the language around the term "flavor".
* Resolving send bounds via RTN: No progress.
* Async drop experiments: @nikomatsakis performed initial review and is continuing to read into the code.
* Async WG reorganizaiton: @tmandry is drafting an initial proposal.

Next steps:

* @nikomatsakis to update Flavor RFC and propose merging once done.
* @compiler-errors to prepare stabilization report for RTN.

**Comment by nikomatsakis** [comment]
Update:

* Upon further discussion the lang team decided to close the flavor RFC (rust-lang/rfcs#3710) and instead to pursue async closure stabilization with the name `AsyncFn`. The premise for the RFC was that we only wanted to commit to `async Fn` if we were ready to commit to at least the *syntax* of `async Trait` being used more broadly for other traits. There were enough questions raised during discussion to give some members pause as to whether we had worked out enough of the design space to commit to that, and so we decided to opt for a less ambitious stabilization instead to give more time for discussion.

**Comment by nikomatsakis** [comment]
With respect to return-type notation, the current status is that we landed an extension of return-type notation to cover `Self::foo(..): Send` in https://github.com/rust-lang/rust/pull/129629 and we are now shaking out the bugs on that. More testing would be helpful, as would reviews on the PRs we've opened already (e.g., https://github.com/rust-lang/rust/issues/132047).

**Comment by nikomatsakis** [comment]
FCP for async closure stabilization [has completed](https://github.com/rust-lang/rust/pull/132706#issuecomment-2506741529). The PR is not yet merged as we are waiting on a reference PR (coming soon), rust-lang/rust#132612 to land, and reviews on the stabilization PR (rust-lang/rust#132706).

**Comment by nikomatsakis** [comment]
Regarding return-type notation, an extension of return-type notation to cover Self::foo(..): Send [landed](https://github.com/rust-lang/rust/pull/129629) and we landed [#132047](https://github.com/rust-lang/rust/issues/132047) which fixes a known ICE. Stabilization PR is now unblocked.

**Comment by nikomatsakis** [comment]
The Dynosaur crate version 0.1.2 was released but some bugs were found post-release. Iteration continues.

**Comment by nikomatsakis** [comment]
Final report of the year: In short, we finished the most important item, and made strong progress on most others but have more work to do:

* Async closures are stabilized (🎉) and will be shipping out on the next nightly. Big kudos to @compiler-errors for driving that.
* Return Type Notation is implemented and we had a call for experimentation but it has not yet reached stable. This will need to be done.
* The work for a standardized macro supporting dynamic dispatch made progress but we haven't yet published the complete story or issued a call for experimentation.

In terms of our stretch goals

* We have had fruitful discussions about the trait for async iteration but do not have widespread consensus, that's on the docker for 2025H1.
* There is a PR with lots of impl work on async closures but we didn't account for reviewing bandwidth. @nikomatsakis has done initial reads and is working with PR author to get this done in 2025H1.

**Comment by nikomatsakis** [comment]
TL;DR:

The biggest news is that Rust 1.85 is stable and includes two major features that impact Async Rust. The first is [async closures](https://blog.rust-lang.org/inside-rust/2024/08/09/async-closures-call-for-testing.html), which has been on many people&#x27;s wish lists for a long time and was expertly moved forward by @compiler-errors over the last year.

The second feature included in 1.85 is the new [lifetime capture rules](https://doc.rust-lang.org/edition-guide/rust-2024/rpit-lifetime-capture.html) as part of Rust 2024 edition. This should substantially improve the experience of using async Rust anytime a user writes &#x60;-&gt; impl Future&#x60;, as it removes the need for &#x60;+ &#x27;_&#x60; or similar bounds in most cases. It will also lead to an easier to understand language, since those bounds only worked by exploiting the more subtle rules of &#x60;impl Trait&#x60; in a way that runs contrary to their actual semantic role in the language. In the 2024 Edition, the subtle rule is gone and we capture all input lifetimes by default, with the ability to use &#x60;+ use&lt;&gt;&#x60; syntax to opt out. See [this blog post](https://blog.rust-lang.org/2024/09/05/impl-trait-capture-rules.html) for more.

**Generators.** The lang team also held a design meeting to review the design for generators, with the outcome of the last one being that we will implement a &#x60;std::iter::iter!&#x60; macro (exact path TBD) in the compiler, as a lang team experiment that allows the use of the &#x60;yield&#x60; syntax. We decided to go in this direction because we want to reserve &#x60;gen&#x60; for self-borrowing and perhaps lending generators, and aren&#x27;t yet decided on which subset of features to expose under that syntax. This decision interacts with ongoing compiler development that isn&#x27;t ready yet to enable experimentation with lending.

Our hope is that in the meantime, by shipping &#x60;iter!&#x60; we will give people the chance to start using generators in their own code and better understand which limitations people hit in practice.

As you may have noticed, I&#x27;m not talking about async generators here. Those are the ultimate goal for the async initiative, but we felt the first step should be clarifying the state of synchronous generators so we can build on that when talking about async ones.

**Dynosaur.** [dynosaur v0.1.3](https://github.com/spastorino/dynosaur/releases/tag/0.1.3) was released, with another release in the works. We think we are approaching a 1.0 release real soon now (tm). At this point you should be able to try it on your crate to enable dyn dispatch for traits with &#x60;async fn&#x60; and other &#x60;-&gt; impl Trait&#x60; methods. If you need to use it together with &#x60;#[trait_variant]&#x60;, you may need to wait until the next release when [#55](https://github.com/spastorino/dynosaur/issues/55) is fixed.

**Comment by tmandry** [comment]
**Generators.** Initial implementation work has started on an `iter!` macro experiment in https://github.com/rust-lang/rust/pull/137725. Discussions have centered around whether the macro should accept blocks in addition to closures, whether thunk closures with an empty arguments list should implement `IntoIterator`, and whether blocks should evaluate to a type that is `Iterator` as well as `IntoIterator`. See the [design meeting notes](https://hackmd.io/iQDQ_J3MTzaKBhq1FTbToQ?view) for more.

**dynosaur.** We released [dynosaur v0.2.0](https://github.com/spastorino/dynosaur/releases/tag/0.2.0) with some critical bug fixes and one breaking change. We have several more breaking changes queued up for an 0.3 release line that we also use plan to use as a 1.0 candidate.

**Pin ergonomics.** https://github.com/rust-lang/rust/pull/135733 landed to implement `&pin const self` and `&pin mut self` sugars as part of the ongoing pin ergonomics experiment. Another PR is open with an early implementation of applying this syntax to borrowing expressions. There has been some discussion within parts of the lang team on whether to prefer this `&pin mut T` syntax or `&mut pin T`, the latter of which applies equally well to `Box<pin T>` but requires an edition.

**Comment by tmandry** [comment]
**Async fn in traits.** An FCP proposal to stabilize return type notation was started in https://github.com/rust-lang/rust/pull/138424. However, it is currently blocked on implementation concerns.

**Async fn in dyn trait.** There have been discussions around next steps to support this in the language. More experimentation is needed, along with an initial RFC.

**dynosaur.** More breaking changes have landed and we expect to release v0.3 soon.

**Comment by tmandry** [comment]
**Generators.** Experimental support for an `iter!` macro has landed in nightly. This is intended for nightly-only experimentation and will still need an RFC before it can stabilize. Tracking issue is [rust-lang/rust#142269](https://github.com/rust-lang/rust/issues/142269).

**Async book.** @nrc has been hard at work filling out the official Async Rust book, recently adding chapters on [concurrency primitives](https://rust-lang.github.io/async-book/part-guide/concurrency-primitives.html), structured concurrency, and pinning.

**dynosaur.** A [dynosaur RFC](https://github.com/spastorino/dynosaur/issues/78) was opened describing what blanket impls we think the proc macro should generate for a trait, to make the trait usable as `impl Trait` in argument position in other traits. This is the last remaining open design question before we release dynosaur 0.3 as a candidate for 1.0. Please chime in on the RFC if you have thoughts.
